src/server/src/main/java/io/cassandrareaper/service/RepairRunner.java

```diff
@@ -344,77 +282,48 @@ private void endRepairRun() {
    */
   private void startNextSegment() throws ReaperException, InterruptedException {
     boolean scheduleRetry = true;
-    boolean anythingRunningStill = false;
 
     // We want to know whether a repair was started,
     // so that a rescheduling of this runner will happen.
     boolean repairStarted = false;
 
-    for (int rangeIndex = 0; rangeIndex < currentlyRunningSegments.length(); rangeIndex++) {
-
-      if (currentlyRunningSegments.get(rangeIndex) != null) {
-        anythingRunningStill = true;
-
-        // Just checking that no currently running segment runner is stuck.
-        RepairSegment supposedlyRunningSegment
-            = context.storage.getRepairSegment(repairRunId, currentlyRunningSegments.get(rangeIndex)).get();
-        DateTime startTime = supposedlyRunningSegment.getStartTime();
-        if (startTime != null && startTime.isBefore(DateTime.now().minusDays(1))) {
-          LOG.warn(
-              "Looks like segment #{} has been running more than a day. Start time: {}",
-              supposedlyRunningSegment.getId(),
-              supposedlyRunningSegment.getStartTime());
-        } else if (startTime != null && startTime.isBefore(DateTime.now().minusHours(1))) {
-          LOG.info(
-              "Looks like segment #{} has been running more than an hour. Start time: {}",
-              supposedlyRunningSegment.getId(),
-              supposedlyRunningSegment.getStartTime());
-        } else if (startTime != null && startTime.isBefore(DateTime.now().minusMinutes(2))) {
-          LOG.debug(
-              "Looks like segment #{} has been running more than two minutes. Start time: {}",
-              supposedlyRunningSegment.getId(),
-              supposedlyRunningSegment.getStartTime());
-        }
-        // No need to try starting new repair for already active slot.
-        continue;
+    // We have an empty slot, so let's start new segment runner if possible.
+    // When in sidecar mode, filter on ranges that the local node is a replica for only.
+    LOG.info("Attempting to run new segment for range");
+    List<RepairSegment> nextRepairSegments
+        = context.config.isInSidecarMode()
+            ? ((IDistributedStorage) context.storage)
+                .getNextFreeSegmentsForRanges(
+                    repairRunId, localEndpointRanges)
+            : context.storage.getNextFreeSegments(
+                repairRunId);
+
+    Optional<RepairSegment> nextRepairSegment = Optional.empty();
+    for (RepairSegment segment : nextRepairSegments) {
+      Collection<String> potentialReplicas = repairUnit.getIncrementalRepair()
+          ? Collections.singletonList(segment.getCoordinatorHost())
+          : segment.getReplicas().keySet();
+      JmxProxy coordinator = clusterFacade.connect(cluster, potentialReplicas);
+      if (nodesReadyForNewRepair(coordinator, segment, segment.getReplicas(), repairRunId)) {
+        nextRepairSegment = Optional.of(segment);
```

Comment:
why keep looping, resetting this each time? 

1. it's possible to work with the first ready  `nextRepairSegment` we found
2. `nodesReadyForNewRepair(..)` involves a round-trip of collecting metrics, i.e. it's blocking on both network calls and remote updates to the metrics table. would this loop be faster it callbacks were in place so the first true response we get from all the `nextRepairSegments` calls to `nodesReadyForNewRepair(..)` is the `nextRepairSegment` to continue with.

Reply:
oops, looks like I forgot a `break;` somewhere 😅  great catch!
Of course, we need to stop the loop as soon as a segment is identified as a valid candidate.
```suggestion
        nextRepairSegment = Optional.of(segment);
        break;
```

Reply:
that addresses (1) but not (2). isn't (2) still relevant/applicable?

Reply:
(2) would involve additional calls potentially and could be a premature optimization. Sending several calls at once would complexify the code and could be fairly heavy (there can be a lot of segments). Keep in mind as well that we may have several Reaper instances doing this concurrently.
The only thing we're checking here really is the number of pending compactions, so there are low chances of having this call returning false.
Also, we're already sending using one thread per node to check on for each segment in the `nodesReadyForNewRepair()` code, so we could end up with a lot of concurrency if we trigger this for 10s or 100s of segments.
Unless we come up with metrics proving that loop is slowing down repairs, I'm -1 on implementing (2).